### PR TITLE
Add a heap profiler API and document it.

### PR DESCRIPTION
**Pull request by hawkinsp**

### CODE REVIEW

File: docs/heap_profiling.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# Heap Profiling
```
Comment (skye) [inline docs/heap_profiling.md:1]: Maybe rename to Device Memory Profiling, I think that'll be clearer to people browsing the docs.

File: docs/heap_profiling.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+pprof --web heap.prof
+```
+
+`pprof` opens a web browser containing the following visualization of the heap
+profile in callgraph format:
+
+![Heap profiling example](_static/heap_profile.svg)
+
+The pprof documentation explains
+[how to interpret callgraph visualizations](https://github.com/google/pprof/blob/master/doc/README.md#interpreting-the-callgraph).
```
Comment (skye) [inline docs/heap_profiling.md:74]: nit: maybe get rid of this, and say something like this at the end of the next paragraph: "For more information on interpreting the callgraph visualizations, see [the pprof documentation](...)"

When I read this, I at first thought this was all that was gonna be said about the visualization.

File: docs/heap_profiling.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+By using `pprof`'s
+[`--diff_base` feature](https://github.com/google/pprof/blob/master/doc/README.md#comparing-profiles) to visualize the change in memory usage
+across loop iterations, we can identify why the memory usage of the
+program increases over time:
+
+```shell
+pprof --web --diff_base heap1.prof heap9.prof
+```
+
+![Heap profile at end of execution](_static/heap_profile_leak2.svg)
```
Comment (skye) [inline docs/heap_profiling.md:140]: Maybe add a tiny bit more explanation, like "Now we see that the memory growth is from the `normal call inside `afunction`."

File: jax/profiler.py [hunk 4]
```diff
@@ -86,3 +87,48 @@ def wrapper(*args, **kwargs):
       return func(*args, **kwargs)
     return wrapper
   return wrapper
+
+
+def heap_profile(backend: Optional[str] = None) -> bytes:
```
Comment (skye) [inline jax/profiler.py:92]: Same here, maybe `device_memory_profile`? (I didn't think of this when reviewing the original XLA changes)

File: docs/jax.profiler.rst [hunk 5]
```diff
@@ -1,6 +1,27 @@
+.. currentmodule:: jax.profiler
+
 jax.profiler module
 ===================
 
 .. automodule:: jax.profiler
-    :members:
-    :show-inheritance:
\ No newline at end of file
+
```
Comment (skye) [inline docs/jax.profiler.rst:7]: Consider linking to the Profiling and Heap Profling pages here (or under the respective titles).

**Review by skye — APPROVED** [review]
Fantastic! Love the illustrations :)

**Comment by hawkinsp** [comment]
I'm going to merge this given Skye has reviewed; if Jake has comments I'm happy to make them in a follow-up PR.

**Pull request merged**